File: Tools/TranscriptedQA/Sources/TranscriptedQA/Utilities/NativeSmokeIsolation.swift
import Darwin
import Foundation

/// HOME overrides do not isolate CFPreferences from the current cfprefsd user.
enum NativeSmokeIsolation {
    static let blockedMessage = "Native Transcripted smoke blocked: this macOS account may share the owner's UserDefaults and TCC state. Run in a separate OS account or a verified hosted CI runner. HOME overrides do not isolate preferences."

    static func allowed(
        uid: uid_t,
        euid: uid_t,
        consoleUID: uid_t?,
        ci: Bool,
        githubActions: Bool,
        githubHosted: Bool,
        virtualMachine: Bool,
        hostedRunnerAccount: Bool
    ) -> Bool {
        guard uid != 0, euid == uid, let consoleUID else { return false }
        if ci && githubActions && githubHosted && (virtualMachine || hostedRunnerAccount) { return true }
        return consoleUID != 0 && uid != consoleUID
    }

    static func isAllowed() -> Bool {
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: "/dev/console"),
              let consoleUID = attributes[.ownerAccountID] as? NSNumber else { return false }
        var virtualMachine: Int32 = 0
        var size = MemoryLayout<Int32>.size
        let hasVM = sysctlbyname("kern.hv_vmm_present", &virtualMachine, &size, nil, 0) == 0
            && virtualMachine == 1
        let environment = ProcessInfo.processInfo.environment
        let account = getpwuid(getuid())
        let hostedRunnerAccount = account.map {
            String(cString: $0.pointee.pw_name) == "runner"
                && String(cString: $0.pointee.pw_dir) == "/Users/runner"
        } ?? false
        return allowed(
            uid: getuid(),
            euid: geteuid(),
            consoleUID: consoleUID.uint32Value,
            ci: environment["CI"] == "true",
            githubActions: environment["GITHUB_ACTIONS"] == "true",
            githubHosted: environment["RUNNER_ENVIRONMENT"] == "github-hosted",
            virtualMachine: hasVM,
            hostedRunnerAccount: hostedRunnerAccount
        )
    }
}

File: scripts/ops/native-smoke-isolation.py
#!/usr/bin/env python3
"""Reject app-launch smokes that could use the owner's macOS preferences.

HOME and CFFIXED_USER_HOME are insufficient: CFPreferences may still resolve
UserDefaults.standard through the real account's cfprefsd session.
"""

from __future__ import annotations

import os
import pwd
import sys


def allowed(
    *,
    uid: int,
    euid: int,
    console_uid: int | None,
    ci: bool,
    github_actions: bool,
    github_hosted: bool,
    hosted_runner_account: bool,
) -> bool:
    if uid == 0 or euid != uid or console_uid is None:
        return False
    if ci and github_actions and github_hosted and hosted_runner_account:
        return True
    return console_uid != 0 and uid != console_uid


def current_state() -> dict[str, object]:
    try:
        console_uid = os.stat("/dev/console").st_uid
    except OSError:
        console_uid = None
    try:
        account = pwd.getpwuid(os.getuid())
        hosted_runner_account = account.pw_name == "runner" and account.pw_dir == "/Users/runner"
    except KeyError:
        hosted_runner_account = False
    return {
        "uid": os.getuid(),
        "euid": os.geteuid(),
        "console_uid": console_uid,
        "ci": os.environ.get("CI") == "true",
        "github_actions": os.environ.get("GITHUB_ACTIONS") == "true",
        "github_hosted": os.environ.get("RUNNER_ENVIRONMENT") == "github-hosted",
        "hosted_runner_account": hosted_runner_account,
    }


def main() -> int:
    if allowed(**current_state()):
        return 0
    print(
        "Native Transcripted smoke blocked: this macOS account may share "
        "the owner's UserDefaults and TCC state. Run in a separate OS account "
        "or a verified hosted CI runner. HOME overrides do not isolate preferences.",
        file=sys.stderr,
    )
    return 3


if __name__ == "__main__":
    raise SystemExit(main())
